### PR TITLE
MCC-1818: Optional authentication support in consensus-service

### PR DESCRIPTION
**Pull request by eranrund**
Soundtrack of this PR: [On and Off the Beaten Path 1](https://soundcloud.com/shoutingfire/oobp1?ref=whatsapp&p=i&c=0)
### Motivation

We'd like to have the option to require authentication when accessing some of our GRPC services. This could allow for things such as centralized rate limiting, or nodes that only accept client requests from specific entities. 

### In this PR

Add support for client request authentication in `mc-consensus-service`. Currently, the only supported authenticated method is a token-based one. A token-generating entity (currently a command line utility, but eventually some service) shares a secret with the token-validating entity (`mc-consensus-service`). The shared secret allows the validating entity to ensure tokens handed to it were signed by a trusted party. Generated tokens contain a unix timestamp of when they were generated, allowing their lifetime to be limited. Other authentication can easily be added in the future.

With regards to code changes, the main ones are:
* A new module, `mc_util_grpc::auth` that contains the authentication primitives.
* A new crate `mc-util-grpc-token-generator` that provides a command line tool for generating tokens, used for testing.
* Two new optional command line arguments to `mc-consensus-service` for enabling token-based authentication (`--client-auth-token-secret` for enabling the authentication, and setting the shared secret and `--client-auth-token-max-lifetime` which allows adjusting the token's max lifetime from the default 24 hours).
* Add username/password support to our `Uri` trait. Those are used to carry authentication information.
* Change attestation GRPC errors to be `PermissionDenied`, and use `Unauthenticated` for authentication errors. We must use separate error codes here since the behavior differs when we encounter them: Upon attestation-related errors the solution is to re-attest (which `ThickClient` can do automatically), but upon authentication error a new token has to be obtained which is outside the scope of `ThickClient`.

While there are some automated tests in place to ensure the basics work, in order to test an actual network with this I did the following:
1) Edit `tools/local-network/local_network.py` and add the required command line argument to consensus service so that it enables token authentication (e.g. `----client-auth-token-secret e9a1cc8b6846eb64248df3cf33afc9c7878ec550f0d221186d577b8ed681218a'
2) Run the local network, `mobilecoind` will fail due to `Unauthenticated`.
3) Generate an authentication token: `cargo run -p mc-util-grpc-token-generator -- --shared-secret e9a1cc8b6846eb64248df3cf33afc9c7878ec550f0d221186d577b8ed681218a --username lol`
4) Edit `tools/local-network/local_network.py` and add the authentication information to the URLs passed to `mobilecoind`: `peers = [f'--peer "insecure-mc://lol:lol%3A1600890302%3Adffd0a03009418cb66e6@localhost:{node.client_port}/"' for node in network.nodes]`. `mobilecoind` now works.

### Future Work
* A standalone service that is capable of handing out authentication tokens based on whatever criteria we come up with (such as rate limits).
* Converting our internal services to support the work presented here.

### CODE REVIEW

File: connection/src/thick.rs [hunk 1]
```diff
@@ -164,7 +170,9 @@ impl AttestedConnection for ThickClient {
         let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
         let (initiator, auth_request_output) = initiator.try_next(&mut csprng, init_input)?;
 
-        let auth_response_msg = self.attested_api_client.auth(&auth_request_output.into())?;
+        let auth_response_msg = self
+            .attested_api_client
+            .auth_opt(&auth_request_output.into(), self.creds.call_option()?)?;
```
Comment (eranrund) [inline connection/src/thick.rs:175]: Passing the `call_option()` returned from `self.creds` is what sticks the authentication HTTP header in the GRPC request.

File: connection/src/traits.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ pub trait AttestedConnection: Connection {
         let result = func(self);
 
         if let Err(GrpcError::RpcFailure(rpc_status)) = &result {
-            if rpc_status.status == RpcStatusCode::UNAUTHENTICATED {
+            if rpc_status.status == RpcStatusCode::PERMISSION_DENIED {
```
Comment (eranrund) [inline connection/src/traits.rs:47]: `UNAUTHENTICATED` is now used for authentication-related stuff, and `PERMISSION_DENIED` is now how we signal clients they need to re-attest.

File: consensus/service/src/attested_api_service.rs [hunk 3]
```diff
@@ -10,31 +10,41 @@ use mc_common::{
...
-use mc_util_grpc::{rpc_logger, rpc_permissions_error, send_result};
+use mc_util_grpc::{auth::Authenticator, rpc_logger, rpc_permissions_error, send_result};
 use mc_util_metrics::SVC_COUNTERS;
 use std::sync::{Arc, Mutex};
 
 #[derive(Clone)]
-pub struct AttestedApiService<E: ConsensusEnclave + Send, S: Session> {
-    enclave: E,
+pub struct AttestedApiService<S: Session> {
+    enclave: Arc<dyn ConsensusEnclave + Send + Sync>,
```
Comment (eranrund) [inline consensus/service/src/attested_api_service.rs:19]: The change to `Arc<dyn>` follows the direction we started taking for the `mockall` tests.

File: consensus/service/src/attested_api_service.rs [hunk 4]
```diff
@@ -10,31 +10,41 @@ use mc_common::{
...
 }
 
-impl<E: ConsensusEnclave + Send> AttestedApi for AttestedApiService<E, PeerSession> {
+impl AttestedApi for AttestedApiService<PeerSession> {
     fn auth(&mut self, ctx: RpcContext, request: AuthMessage, sink: UnarySink<AuthMessage>) {
         let _timer = SVC_COUNTERS.req(&ctx);
         mc_common::logger::scoped_global_logger(&rpc_logger(&ctx, &self.logger), |logger| {
+            if let Err(err) = self.authenticator.authenticate_rpc(&ctx) {
+                return send_result(ctx, sink, err.into(), &logger);
+            }
```
Comment (eranrund) [inline consensus/service/src/attested_api_service.rs:46]: This is the pattern we need to follow when we want to require RPC calls to go through authentication.

File: consensus/service/src/attested_api_service.rs [hunk 5]
```diff
@@ -10,31 +10,41 @@ use mc_common::{
...
     logger: Logger,
     sessions: Arc<Mutex<HashSet<S>>>,
 }
 
-impl<E: ConsensusEnclave + Send, S: Session> AttestedApiService<E, S> {
-    pub fn new(enclave: E, logger: Logger) -> Self {
+impl<S: Session> AttestedApiService<S> {
+    pub fn new(
+        enclave: Arc<dyn ConsensusEnclave + Send + Sync>,
+        authenticator: Arc<dyn Authenticator + Send + Sync>,
```
Comment (eranrund) [inline consensus/service/src/attested_api_service.rs:28]: This is the plugable authentication mechanism.

File: consensus/service/src/blockchain_api_service.rs [hunk 6]
```diff
@@ -141,14 +157,45 @@ impl<L: Ledger + Clone> BlockchainApi for BlockchainApiService<L> {
...
+        let ch = ChannelBuilder::new(env).connect(&format!("127.0.0.1:{}", port));
+        let client = BlockchainApiClient::new(ch);
+        (client, server)
+    }
 
     #[test_with_logger]
     // `get_last_block_info` should returns the last block.
     fn test_get_last_block_info(logger: Logger) {
         let mut ledger_db = create_ledger();
+        let authenticator = Arc::new(AnonymousAuthenticator::default());
```
Comment (eranrund) [inline consensus/service/src/blockchain_api_service.rs:198]: The `AnonymousAuthenticator` allows all requests to go through.

File: consensus/service/src/consensus_service.rs [hunk 7]
```diff
@@ -396,13 +421,14 @@ impl<
...
                 self.ledger_db.clone(),
+                peer_authenticator.clone(),
                 self.logger.clone(),
             ),
         );
 
         let peer_service =
             consensus_peer_grpc::create_consensus_peer_api(peer_api_service::PeerApiService::new(
-                self.enclave.clone(),
+                self.enclave.clone(), // TODO use Arc'ed enclave from above
```
Comment (eranrund) [inline consensus/service/src/consensus_service.rs:431]: I'm filing a ticket to address this.

**Comment by mfaulk** [comment]
This is looking good, although in the long term we should check out Tonic for gRPC. It looks like it provides (or, at least, provides a pattern for implementing) a lot of good things. For example, adding an "interceptor" to outgoing/incoming requests which adds/checks authentication data:

https://github.com/hyperium/tonic/blob/master/examples/src/authentication/client.rs#L14
https://github.com/hyperium/tonic/blob/master/examples/src/authentication/server.rs#L54

This is pretty nice because it allows layering on authentication without modifying the base client/server functionality.

**Comment by eranrund** [comment]
> This is looking good, although in the long term we should check out Tonic for gRPC. It looks like it provides (or, at least, provides a pattern for implementing) a lot of good things. For example, adding an "interceptor" to outgoing/incoming requests which adds/checks authentication data:
> 
> https://github.com/hyperium/tonic/blob/master/examples/src/authentication/client.rs#L14
> https://github.com/hyperium/tonic/blob/master/examples/src/authentication/server.rs#L54
> 
> This is pretty nice because it allows layering on authentication without modifying the base client/server functionality.

Nice, I'll be happy to look into it at some point.

**Review by mfaulk — APPROVED** [review]

File: consensus/service/src/config.rs [hunk 8]
```diff
@@ -94,6 +107,26 @@ fn keypair_from_base64(private_key: &str) -> Result<Arc<Ed25519Pair>, String> {
     Ok(Arc::new(Ed25519Pair::from(secret_key)))
 }
 
+/// Converts a hex-encoded string into an array of 32 bytes.
+fn from_hex_32(src: &str) -> Result<[u8; 32], String> {
+    let bytes = hex::decode(src).map_err(|err| format!("Invalid input: {}", err))?;
```
Comment (jcape) [inline consensus/service/src/config.rs:112]: Recommend check `src.len()` and using [`hex::decode_to_slice()`](https://docs.rs/hex/0.4.2/hex/fn.decode_to_slice.html) instead.

File: util/grpc-token-generator/src/bin/main.rs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    println!("Password: {}", creds.password());
+    println!(
+        "Password (percent-encoded): {}",
+        utf8_percent_encode(creds.password(), NON_ALPHANUMERIC).to_string()
+    );
+}
+
+/// Converts a hex-encoded string into an array of 32 bytes.
+fn from_hex_32(src: &str) -> Result<[u8; 32], String> {
+    let bytes = hex::decode(src).map_err(|err| format!("Invalid input: {}", err))?;
```
Comment (jcape) [inline util/grpc-token-generator/src/bin/main.rs:40]: Same story as the other option parser.

File: util/grpc/src/auth/mod.rs [hunk 10]
```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! GRPC authentication utilities.
+
+pub mod anonymous_authenticator;
+pub mod token_authenticator;
```
Comment (jcape) [inline util/grpc/src/auth/mod.rs:6]: Recommend not making these public unless there's a reason to do so.

File: util/grpc/src/auth/token_authenticator.rs [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    fn is_valid_signature(&self, data: &str, signature: &str) -> Result<bool, AuthenticatorError> {
+        let mut mac = Hmac::<sha2::Sha256>::new_varkey(&self.shared_secret)
+            .map_err(|_| AuthenticatorError::Other("Invalid HMAC key".to_owned()))?;
+        mac.update(data.as_bytes());
+        let our_signature = mac.finalize().into_bytes();
+
+        let their_suffix: Vec<u8> =
+            hex::decode(signature).map_err(|_| AuthenticatorError::InvalidAuthorizationToken)?;
```
Comment (jcape) [inline util/grpc/src/auth/token_authenticator.rs:95]: This can probably go at the top of this function, as signal is doing it. I don't think it precludes timing attacks against `new_varkey()`, if there are any, but it probably won't hurt to validate user input before accessing secrets... ;)

File: util/grpc/src/auth/token_authenticator.rs [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+impl TokenBasicCredentialsGenerator {
+    pub fn new(shared_secret: [u8; 32]) -> Self {
+        Self { shared_secret }
+    }
+
+    pub fn generate_for(
+        &self,
+        user_id: &str,
+    ) -> Result<BasicCredentials, TokenBasicCredentialsGeneratorError> {
+        let current_time_seconds = SystemTime::now()
```
Comment (jcape) [inline util/grpc/src/auth/token_authenticator.rs:126]: This shows up in various places, and it's probably OK, but one thing to consider is adding a `TimeProvider::from_epoch()` trait / method, and implementing it in terms of `SystemTime`, then making these things all generic over that (e.g. `TokenThing<TimeProvider = SystemTime>`).

At minimum, this will allow mocking out the source of "now" to make all tests deterministic, as well as let us test weird values and bounds conditions.

**Review by jcape — COMMENTED** [review]
Mostly looks good to me, some things can be improved (most notably, making "now" abstract to allow better testing of edge cases).

File: common/src/lib.rs [hunk 13]
```diff
@@ -53,3 +53,10 @@ cfg_if::cfg_if! {
         pub mod sentry;
     }
 }
+
+// Time
+cfg_if::cfg_if! {
+    if #[cfg(feature = "std")] {
+        pub mod time;
+    }
+}
```
Comment (jcape) [inline common/src/lib.rs:62]: Suggest making this module public in all cases, and only gating the `TimeProvider` implementation on `SystemTime` on the `std` feature.

**Review by jcape — APPROVED** [review]
LGTM as-is, though you may want to make TimeProvider no-std and just gate the SystemTime implementation of it behind the feature...

File: common/src/lib.rs (hunk 13 above)
Comment (eranrund) [reply common/src/lib.rs:62]: Sounds good, will do and then merge.
